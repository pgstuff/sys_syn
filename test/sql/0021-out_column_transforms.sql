BEGIN;

CREATE EXTENSION tinyint SCHEMA public;
CREATE EXTENSION sys_syn;

CREATE SCHEMA user_data
    AUTHORIZATION postgres;

CREATE TABLE user_data.test_table (
        test_table_id integer NOT NULL,
        test_table_text text,
        test_table_date date,
        test_table_datetime timestamp with time zone,
        CONSTRAINT test_table_pid PRIMARY KEY (test_table_id));

INSERT INTO sys_syn.in_groups_def VALUES ('in');

SELECT sys_syn.in_table_create_sql('user_data.test_table'::regclass, 'in');

DO $$BEGIN
    EXECUTE sys_syn.in_table_create_sql('user_data.test_table'::regclass, 'in');
END$$;

INSERT INTO user_data.test_table (
        test_table_id, test_table_text,        test_table_date,        test_table_datetime)
VALUES (1,              'test_record_1',        '2010-01-02',           '2013-04-05 06:07:08-00');
INSERT INTO user_data.test_table (
        test_table_id, test_table_text,        test_table_date,        test_table_datetime)
VALUES (2,              '',                     'infinity'::DATE,       'infinity'::timestamp);
INSERT INTO user_data.test_table (
        test_table_id, test_table_text,        test_table_date,        test_table_datetime)
VALUES (3,              NULL,                   '-infinity'::DATE,      '-infinity'::timestamp);

INSERT INTO sys_syn.out_column_transforms(
        rule_group_id,                priority,       data_type_like,                 in_table_id_like,       out_group_id_like,
        column_name_like,       in_column_type,         new_data_type,
        expression,
        omit,   new_column_name,        final_rule)
VALUES ('time_zone_us_eastern', 50,          'timestamp with time zone',  NULL,                   NULL,
        NULL,                   NULL,                   'timestamp without time zone',
        $$%1 AT TIME ZONE 'US/Eastern'$$,
        NULL,   NULL,                   FALSE);

INSERT INTO sys_syn.out_column_transforms(
        rule_group_id,                priority,       data_type_like,                 in_table_id_like,       out_group_id_like,
        column_name_like,       in_column_type,         new_data_type,
        expression,
        omit,   new_column_name,        final_rule)
VALUES (NULL,                   10,             NULL,  NULL,                   NULL,
        'sys_syn_trans_id_out', NULL,                   NULL,
        NULL,
        TRUE,   NULL,                   TRUE);

INSERT INTO sys_syn.out_column_transforms(
        rule_group_id,          priority,       final_ids,              data_type_like,                 in_table_id_like,
        out_group_id_like,      column_name_like,       in_column_type,         new_data_type,
        expression,
        omit,   new_column_name,        final_rule)
VALUES (NULL,                   101,            '{date_infinity}',      'date',                         NULL,
        NULL,                   NULL,                   NULL,                   NULL,
        $$final_id date_infinity not set!$$,
        NULL,   NULL,                   FALSE);

INSERT INTO sys_syn.out_groups_def (
        out_group_id,   out_column_transform_rule_group_ids)
VALUES ('out',          ARRAY['time_zone_us_eastern','sys_syn-mssql']);

SELECT sys_syn.out_table_create_sql('user_data', 'test_table', 'out', omit_columns => ARRAY['sys_syn_hold_reason_text'], data_view => TRUE);

DO $$BEGIN
        EXECUTE sys_syn.out_table_create_sql('user_data', 'test_table', 'out', omit_columns => ARRAY['sys_syn_hold_reason_text'], data_view => TRUE);
END$$;

SELECT user_data.test_table_pull(FALSE);
SELECT user_data.test_table_out_move_1();

SELECT * FROM user_data.test_table_out_queue_data_1;

UPDATE user_data.test_table_out_queue_data_1 SET sys_syn_queue_state = 'Claimed'::sys_syn.queue_state WHERE test_table_id = 1;

UPDATE user_data.test_table_out_queue_data_1 SET sys_syn_queue_state = 'Processed'::sys_syn.queue_state WHERE test_table_id = 1;

SELECT user_data.test_table_out_processed_1();

SELECT * FROM user_data.test_table_out_queue_data_1;

ROLLBACK;
