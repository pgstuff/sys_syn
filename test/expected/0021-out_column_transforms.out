BEGIN;
CREATE EXTENSION tinyint
    SCHEMA public;
CREATE EXTENSION sys_syn;
CREATE SCHEMA user_data
    AUTHORIZATION postgres;
CREATE TABLE user_data.test_table (
        test_table_id integer NOT NULL,
        test_table_text text,
        test_table_date date,
        test_table_datetime timestamp with time zone,
        CONSTRAINT test_table_pid PRIMARY KEY (test_table_id));
INSERT INTO sys_syn.in_groups_def VALUES ('in');
SELECT sys_syn.in_table_add_sql('user_data.test_table'::regclass, 'in');
                                                         in_table_add_sql                                                          
-----------------------------------------------------------------------------------------------------------------------------------
 SELECT sys_syn.in_table_add (                                                                                                    +
                 schema          => 'user_data'::regnamespace,                                                                    +
                 in_table_id     => 'test_table',                                                                                 +
                 in_group_id     => 'in',                                                                                         +
                 in_pull_id      => NULL,                                                                                         +
                 in_columns      => ARRAY[                                                                                        +
                        $COL$("test_table_id","integer",Id,"in_source.test_table_id",,,,)$COL$,                                   +
                        $COL$("test_table_text","text",Attribute,"in_source.test_table_text",,,,)$COL$,                           +
                        $COL$("test_table_date","date",Attribute,"in_source.test_table_date",,,,)$COL$,                           +
                        $COL$("test_table_datetime","timestamp with time zone",Attribute,"in_source.test_table_datetime",,,,)$COL$+
                 ]::sys_syn.create_in_column[],                                                                                   +
                 full_table_reference    => 'user_data.test_table',                                                               +
                 changes_table_reference => NULL,                                                                                 +
                 full_sql                => NULL,                                                                                 +
                 changes_sql             => NULL,                                                                                 +
                 full_pre_sql            => NULL,                                                                                 +
                 changes_pre_sql         => NULL,                                                                                 +
                 full_post_sql           => NULL,                                                                                 +
                 changes_post_sql        => NULL,                                                                                 +
                 enable_deletes_implied  => TRUE,                                                                                 +
                 full_prepull_id         => NULL,                                                                                 +
                 changes_prepull_id      => NULL                                                                                  +
         );
(1 row)

DO $$BEGIN
    EXECUTE sys_syn.in_table_add_sql('user_data.test_table'::regclass, 'in');
END$$;
INSERT INTO user_data.test_table (
        test_table_id, test_table_text,        test_table_date,        test_table_datetime)
VALUES (1,              'test_record_1',        '2010-01-02',           '2013-04-05 06:07:08-00');
INSERT INTO user_data.test_table (
        test_table_id, test_table_text,        test_table_date,        test_table_datetime)
VALUES (2,              '',                     'infinity'::DATE,       'infinity'::timestamp);
INSERT INTO user_data.test_table (
        test_table_id, test_table_text,        test_table_date,        test_table_datetime)
VALUES (3,              NULL,                   '-infinity'::DATE,      '-infinity'::timestamp);
INSERT INTO sys_syn.out_column_transforms(
        rule_group_id,                priority,       data_type_like,                 in_table_id_like,       out_group_id_like,
        column_name_like,       in_column_type,         new_data_type,
        expression,
        omit,   new_column_name,        final_rule)
VALUES ('time_zone_us_eastern', 50,          'timestamp with time zone',  NULL,                   NULL,
        NULL,                   NULL,                   'timestamp without time zone',
        $$%1 AT TIME ZONE 'US/Eastern'$$,
        NULL,   NULL,                   FALSE);
INSERT INTO sys_syn.out_column_transforms(
        rule_group_id,                priority,       data_type_like,                 in_table_id_like,       out_group_id_like,
        column_name_like,       in_column_type,         new_data_type,
        expression,
        omit,   new_column_name,        final_rule)
VALUES (NULL,                   10,             NULL,  NULL,                   NULL,
        'sys_syn_trans_id_out', NULL,                   NULL,
        NULL,
        TRUE,   NULL,                   TRUE);
INSERT INTO sys_syn.out_column_transforms(
        rule_group_id,          priority,       final_ids,              data_type_like,                 in_table_id_like,
        out_group_id_like,      column_name_like,       in_column_type,         new_data_type,
        expression,
        omit,   new_column_name,        final_rule)
VALUES (NULL,                   101,            '{date_infinity}',      'date',                         NULL,
        NULL,                   NULL,                   NULL,                   NULL,
        $$final_id date_infinity not set!$$,
        NULL,   NULL,                   FALSE);
INSERT INTO sys_syn.out_groups_def (
        out_group_id,   out_column_transform_rule_group_ids)
VALUES ('out',          ARRAY['time_zone_us_eastern','sys_syn-mssql']);
SELECT sys_syn.out_table_add_sql('user_data', 'test_table', 'out', omit_columns => ARRAY['sys_syn_hold_reason_text'], data_view => TRUE);
                                                                                                                                                                                                                                                                         out_table_add_sql                                                                                                                                                                                                                                                                         
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 SELECT sys_syn.out_table_add (                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   +
                 schema                  => 'user_data'::regnamespace,                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            +
                 in_table_id             => 'test_table',                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         +
                 out_group_id            => 'out',                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                +
                 out_columns             => ARRAY[                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                +
                        $COL$("sys_syn_trans_id_in","out_queue.trans_id_in",,)$COL$,                                                                                                                                                                                                                                                                                                                                                                                                                                                                              +
                        $COL$("sys_syn_delta_type","out_queue.delta_type",,)$COL$,                                                                                                                                                                                                                                                                                                                                                                                                                                                                                +
                        $COL$("sys_syn_queue_state","out_queue.queue_state",queue_state,"new.sys_syn_queue_state")$COL$,                                                                                                                                                                                                                                                                                                                                                                                                                                          +
                        $COL$("sys_syn_queue_id","out_queue.queue_id",queue_id,"new.sys_syn_queue_id")$COL$,                                                                                                                                                                                                                                                                                                                                                                                                                                                      +
                        $COL$("sys_syn_queue_priority","out_queue.queue_priority",queue_priority,"new.sys_syn_queue_priority")$COL$,                                                                                                                                                                                                                                                                                                                                                                                                                              +
                        $COL$("sys_syn_hold_updated","out_queue.hold_updated",,)$COL$,                                                                                                                                                                                                                                                                                                                                                                                                                                                                            +
                        $COL$("sys_syn_hold_trans_id_first","out_queue.hold_trans_id_first",,)$COL$,                                                                                                                                                                                                                                                                                                                                                                                                                                                              +
                        $COL$("sys_syn_hold_trans_id_last","out_queue.hold_trans_id_last",,)$COL$,                                                                                                                                                                                                                                                                                                                                                                                                                                                                +
                        $COL$("sys_syn_hold_reason_count","out_queue.hold_reason_count",,)$COL$,                                                                                                                                                                                                                                                                                                                                                                                                                                                                  +
                        $COL$("sys_syn_hold_reason_id","out_queue.hold_reason_id",hold_reason_id,"new.sys_syn_hold_reason_id")$COL$,                                                                                                                                                                                                                                                                                                                                                                                                                              +
                        $COL$("sys_syn_processed_time","out_queue.processed_time",processed_time,"new.sys_syn_processed_time")$COL$,                                                                                                                                                                                                                                                                                                                                                                                                                              +
                        $COL$("test_table_id","(in_source.id).test_table_id",,)$COL$,                                                                                                                                                                                                                                                                                                                                                                                                                                                                             +
                        $COL$("test_table_text","(in_source.attributes).test_table_text",,)$COL$,                                                                                                                                                                                                                                                                                                                                                                                                                                                                 +
                        $COL$("test_table_date","CASE WHEN (in_source.attributes).test_table_date < '0001-01-01'::DATE THEN '0001-01-01'::DATE WHEN (in_source.attributes).test_table_date > '9999-12-31'::DATE THEN '9999-12-31'::DATE ELSE (in_source.attributes).test_table_date END",,)$COL$,                                                                                                                                                                                                                                                                 +
                        $COL$("test_table_datetime","CASE WHEN (in_source.attributes).test_table_datetime AT TIME ZONE 'US/Eastern' < '1753-01-01 00:00:00'::timestamp without time zone THEN '1753-01-01 00:00:00'::timestamp without time zone WHEN (in_source.attributes).test_table_datetime AT TIME ZONE 'US/Eastern' > '9999-12-31 23:59:59.997'::timestamp without time zone THEN '9999-12-31 23:59:59.997'::timestamp without time zone ELSE date_trunc('milliseconds', (in_source.attributes).test_table_datetime AT TIME ZONE 'US/Eastern') END",,)$COL$+
                 ]::sys_syn.create_out_column[],                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  +
                 data_view               => 'true',                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               +
                 out_log_lifetime        => NULL,                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 +
                 notification_channel    => NULL,                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 +
                 enable_adds             => 'true',                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               +
                 enable_changes          => 'true',                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               +
                 enable_deletes          => 'true',                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               +
                 condition_sql           => NULL,                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 +
                 claim_limit_rows        => '2147483647',                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         +
                 claim_queue_count       => NULL,                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 +
                 claim_fixed_by_id       => 'false',                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              +
                 claim_random_sample     => NULL,                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 +
                 queue_pid_used_age      => NULL,                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 +
                 record_comparison_different=> NULL,                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              +
                 record_comparison_same     => NULL                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               +
         );
(1 row)

DO $$BEGIN
    EXECUTE sys_syn.out_table_add_sql('user_data', 'test_table', 'out', omit_columns => ARRAY['sys_syn_hold_reason_text'], data_view => TRUE);
END$$;
SELECT user_data.test_table_pull(FALSE);
 test_table_pull 
-----------------
 t
(1 row)

SELECT user_data.test_table_out_move();
 test_table_out_move 
---------------------
 t
(1 row)

SELECT * FROM user_data.test_table_out_queue_data;
 sys_syn_id | sys_syn_trans_id_in | sys_syn_delta_type | sys_syn_queue_state | sys_syn_queue_id | sys_syn_queue_priority | sys_syn_hold_updated | sys_syn_hold_trans_id_first | sys_syn_hold_trans_id_last | sys_syn_hold_reason_count | sys_syn_hold_reason_id | sys_syn_processed_time | test_table_id | test_table_text | test_table_date |     test_table_datetime      
------------+---------------------+--------------------+---------------------+------------------+------------------------+----------------------+-----------------------------+----------------------------+---------------------------+------------------------+------------------------+---------------+-----------------+-----------------+------------------------------
 (1)        |                   1 | Add                | Unclaimed           |                  |                        |                      |                             |                            |                           |                        |                        |             1 | test_record_1   | 01-02-2010      | Fri Apr 05 02:07:08 2013
 (2)        |                   1 | Add                | Unclaimed           |                  |                        |                      |                             |                            |                           |                        |                        |             2 |                 | 12-31-9999      | Fri Dec 31 23:59:59.997 9999
 (3)        |                   1 | Add                | Unclaimed           |                  |                        |                      |                             |                            |                           |                        |                        |             3 |                 | 01-01-0001      | Mon Jan 01 00:00:00 1753
(3 rows)

UPDATE user_data.test_table_out_queue_data SET sys_syn_queue_state = 'Claimed'::sys_syn.queue_state WHERE test_table_id = 1;
UPDATE user_data.test_table_out_queue_data SET sys_syn_queue_state = 'Processed'::sys_syn.queue_state WHERE test_table_id = 1;
SELECT user_data.test_table_out_processed();
 test_table_out_processed 
--------------------------
 t
(1 row)

SELECT * FROM user_data.test_table_out_queue_data;
 sys_syn_id | sys_syn_trans_id_in | sys_syn_delta_type | sys_syn_queue_state | sys_syn_queue_id | sys_syn_queue_priority | sys_syn_hold_updated | sys_syn_hold_trans_id_first | sys_syn_hold_trans_id_last | sys_syn_hold_reason_count | sys_syn_hold_reason_id | sys_syn_processed_time | test_table_id | test_table_text | test_table_date |     test_table_datetime      
------------+---------------------+--------------------+---------------------+------------------+------------------------+----------------------+-----------------------------+----------------------------+---------------------------+------------------------+------------------------+---------------+-----------------+-----------------+------------------------------
 (2)        |                   1 | Add                | Unclaimed           |                  |                        |                      |                             |                            |                           |                        |                        |             2 |                 | 12-31-9999      | Fri Dec 31 23:59:59.997 9999
 (3)        |                   1 | Add                | Unclaimed           |                  |                        |                      |                             |                            |                           |                        |                        |             3 |                 | 01-01-0001      | Mon Jan 01 00:00:00 1753
(2 rows)

ROLLBACK;
